src/index/sparse/sparse_inverted_index.h

```diff
@@ -561,74 +565,77 @@ class InvertedIndex : public BaseInvertedIndex<T> {
 
     // LUT supports size() and operator[] which returns an SparseIdVal.
     template <typename LUT>
-    class Cursor {
+    struct Cursor {
      public:
         Cursor(const LUT& lut, size_t num_vec, float max_score, float q_value, const BitsetView bitset)
-            : lut_(lut), num_vec_(num_vec), max_score_(max_score), q_value_(q_value), bitset_(bitset) {
-            while (loc_ < lut_.size() && !bitset_.empty() && bitset_.test(cur_vec_id())) {
+            : lut_(lut),
+              lut_size_(lut.size()),
+              total_num_vec_(num_vec),
+              max_score_(max_score),
+              q_value_(q_value),
+              bitset_(bitset) {
+            while (loc_ < lut_size_ && !bitset_.empty() && bitset_.test(lut_[loc_].id)) {
                 loc_++;
             }
+            update_cur_vec_id();
         }
         Cursor(const Cursor& rhs) = delete;
 
         void
         next() {
-            loc_++;
-            while (loc_ < lut_.size() && !bitset_.empty() && bitset_.test(cur_vec_id())) {
-                loc_++;
-            }
+            next_internal();
+            update_cur_vec_id();
         }
-        // advance loc until cur_vec_id() >= vec_id
+
+        // advance loc until cur_vec_id_ >= vec_id
         void
         seek(table_t vec_id) {
-            while (loc_ < lut_.size() && cur_vec_id() < vec_id) {
-                next();
+            while (loc_ < lut_size_ && lut_[loc_].id < vec_id) {
+                next_internal();
             }
+            update_cur_vec_id();
         }
-        [[nodiscard]] table_t
-        cur_vec_id() const {
-            if (is_end()) {
-                return num_vec_;
-            }
-            return lut_[loc_].id;
-        }
+
         T
         cur_vec_val() const {
             return lut_[loc_].val;
         }
-        [[nodiscard]] bool
-        is_end() const {
-            return loc_ >= size();
-        }
-        [[nodiscard]] float
-        q_value() const {
-            return q_value_;
-        }
-        [[nodiscard]] size_t
-        size() const {
-            return lut_.size();
-        }
-        [[nodiscard]] float
-        max_score() const {
-            return max_score_;
-        }
 
-     private:
         const LUT& lut_;
+        const size_t lut_size_;
         size_t loc_ = 0;
-        size_t num_vec_ = 0;
+        size_t total_num_vec_ = 0;
         float max_score_ = 0.0f;
         float q_value_ = 0.0f;
         const BitsetView bitset_;
-    };  // class Cursor
+        table_t cur_vec_id_ = 0;
+
+     private:
+        inline void
+        update_cur_vec_id() {
+            if (loc_ >= lut_size_) {
+                cur_vec_id_ = total_num_vec_;
```

Comment:
just double-checking that this is not `total_num_vec_ - 1`. According to standard naming conventions, `id_` is assumed to be something within `[0, num)` range.

Reply:
yes it is not `total_num_vec_ - 1`. When `id_` is in `[0, num)` range it is still in range and the cursor is pointing at some doc, `loc_ >= lut_size_` means it is out of range and the cursor points at no doc.